### PR TITLE
fix(anvil): fix incorrect op-stack deposit tx hashes 

### PR DESCRIPTION
**Pull request by jakim929**
## Motivation
tx hash calculation for deposit tx for `anvil --optimism` was different from [specs](https://specs.optimism.io/protocol/deposits.html#the-deposited-transaction-type). this made it difficult to deterministically generate the L2 tx hash from the L1 deposit transaction.

- the EIP-2718 tx type was not included when calculating hash
- encoding was slightly different from [spec](https://specs.optimism.io/protocol/deposits.html#the-deposited-transaction-type) 
    - nonce was included in the rlp encoding
    - order of fields was different from spec

## Solution
- calculate tx hash after encoding in eip-2718 envelope format
- fix order of fields / remove nonce when RLP encoding / decoding
- added tests - checks against a op-sepolia tx hash / fields to make sure that the hash / encoding is consistent

### Test to repro issue
the following test would fail previously
```rust
    fn test_decode_encode_deposit_tx() {
        // https://sepolia-optimism.etherscan.io/tx/0xbf8b5f08c43e4b860715cd64fc0849bbce0d0ea20a76b269e7bc8886d112fca7
        let tx_hash: TxHash = "0xbf8b5f08c43e4b860715cd64fc0849bbce0d0ea20a76b269e7bc8886d112fca7"
            .parse::<TxHash>()
            .unwrap();

        // https://sepolia-optimism.etherscan.io/getRawTx?tx=0xbf8b5f08c43e4b860715cd64fc0849bbce0d0ea20a76b269e7bc8886d112fca7
        let raw_tx = alloy_primitives::hex::decode(
           "7ef861a0dfd7ae78bf3c414cfaa77f13c0205c82eb9365e217b2daa3448c3156b69b27ac94778f2146f48179643473b82931c4cd7b8f153efd94778f2146f48179643473b82931c4cd7b8f153efd872386f26fc10000872386f26fc10000830186a08080",
        )
        .unwrap();
        let dep_tx = TypedTransaction::decode(&mut raw_tx.as_slice()).unwrap();
        let mut encoded = Vec::<u8>::new();
        dep_tx.encode_2718(&mut encoded);
        assert_eq!(raw_tx, encoded);
        assert_eq!(tx_hash, dep_tx.hash());
    }
```

### CODE REVIEW

**Review by mattsse — CHANGES_REQUESTED** [review]
this is great, ty

few pedantic nits

we now have this type also here  https://github.com/alloy-rs/op-alloy so we could replace the entire anvil deposit type, will track this separately.

q, are you using eth_sendraw deposit txs at OP for testing?

File: crates/anvil/core/src/eth/transaction/optimism.rs [hunk 1]
```diff
@@ -20,13 +23,17 @@ pub struct DepositTransactionRequest {
 
 impl DepositTransactionRequest {
     pub fn hash(&self) -> B256 {
-        B256::from_slice(alloy_primitives::keccak256(alloy_rlp::encode(self)).as_slice())
+        let mut encoded = Vec::<u8>::new();
```
[the remark this answers is not in the document]
Comment (jakim929) [reply crates/anvil/core/src/eth/transaction/optimism.rs]: will fix!

File: crates/anvil/core/src/eth/transaction/optimism.rs [hunk 2]
```diff
@@ -286,6 +299,13 @@ impl DepositTransaction {
         len
     }
 
+    pub fn decode_2718(buf: &mut &[u8]) -> Result<Self, DecodeError> {
+        use bytes::Buf;
+        // Skip the tx type byte
+        buf.advance(1);
```
[the remark this answers is not in the document]
Comment (jakim929) [reply crates/anvil/core/src/eth/transaction/optimism.rs:312]: added!

**Comment by jakim929** [comment]
thanks for the quick review! 🙏  addressed the requested changes

> q, are you using eth_sendraw deposit txs at OP for testing?

Yep, it's very useful for testing bridging! and we're using it pretty extensively for https://github.com/ethereum-optimism/supersim , which is a local environment for testing L1 + multiple L2s (anvils) for bridging, message passing etc.

Also we're happy to help maintain the optimism related features on anvil including tasks like
> we now have this type also here https://github.com/alloy-rs/op-alloy so we could replace the entire anvil deposit type, will track this separately.

and improving support for deposit tx validation in the tx pool.

**Review by mattsse — APPROVED** [review]

**Pull request merged**